README.md
TorTP
==========

TorTP will change your iptables configuration to force all TCP traffic to pass through Tor (and also UDP DNS request). The network traffic that is not capable of passing through Tor (such as UDP or ICMP) is just dropped.

TorTP use python-stem library for setup Transparen Proxy and DNS server capability on Tor, without override Tor default configuration file.

HowTo install:
=============

Add freepto repository:

<code>$ wget http://deb.freepto.mx/deb.gpg</code>

<code>$ cat deb.gpg | apt-key add -</code>

<code>$ echo "deb http://deb.freepto.mx/freeptorepo berenjena main" > /etc/apt/sources.list.d/freepto.list</code>

Install:

<code>sudo apt-get install tortp python-tortp</code>

Enable Tor Control Port:

<code>sed -i 's/#ControlPort 9051/ControlPort 9051/' /etc/tor/torrc</code>

Install GUI (optional):

<code>apt-get install tortp-gtk</code>

HowTo user TorTP:
=================

<code>$ sudo tortp -h</code>

<code>$ man tortp</code>


debian/changelog
tortp (0.5) wheezy; urgency=medium

  * Bump version after debian packaging

 -- freepto devs <freepto@autistici.org>  Fri, 10 Jul 2015 23:23:51 +0200

tortp (0.4) wheezy; urgency=low

  * Initial release.

 -- freepto devs <freepto@autistici.org>  Mon, 22 Jun 2015 15:02:09 +0200


tort.py
#!/usr/local/bin/python

"""
TORtp is a simple way to implement
Tor Transparent Proxy in our GNU/Linux Box

TORtp use TOR's control port for setup
Transparen Proxy and DNS server capability
on TOR, without override default configuration file
with a custom torrc
"""

import argparse
import subprocess
import os
import pynotify
import stem
from stem.control import Controller
import stem.util
import stem.process
from stem import CircStatus
from shutil import copy2
import sys

def notify(title, message):
    pynotify.init("TORtp")
    notice = pynotify.Notification(title, message, "/usr/share/icons/tortp/anonymous.ico")
    notice.show()
    return

def check_user():
   """ Only root can do that! """
   uid = subprocess.Popen('id -u', shell=True, stdout = subprocess.PIPE)
   out = uid.stdout.read()
   if int(out) == 0:
      return os.environ['SUDO_UID']
   else:
      notify("TORtp", "Only root can do that!")
      sys.exit(1)

def set_home(user):
   """ Set user home path"""
   h = subprocess.Popen('egrep "%s:" /etc/passwd | cut -d : -f 6' % user, shell=True, stdout = subprocess.PIPE)
   home = h.stdout.read()
   return home.strip()

def tortp_dir(home):
   """ Create /home/$user/.tortp """
   tortpdir = "%s/.tortp" % home
   if not os.path.exists(tortpdir):
      os.makedirs(tortpdir)
      notify("TORtp", "Directory %s created" % tortpdir)
   return tortpdir

def check_sys_dependecies():
   """ check if all dependencies are installed """
   devnull = open(os.devnull,"w")
   dnsmasq = subprocess.call(["dpkg","-s","dnsmasq"],stdout=devnull,stderr=subprocess.STDOUT)
   if dnsmasq != 0:
      print "Package dnsmasq not installed"
   tor = subprocess.call(["dpkg","-s","tor"],stdout=devnull,stderr=subprocess.STDOUT)
   if tor != 0:
       print "Package tor not installed"
   devnull.close()

def iptables_clean():
   """ This function remove all iptables rules """
   subprocess.call('iptables -F', shell=True)
   subprocess.call('iptables -X', shell=True)
   subprocess.call('iptables -t nat -F', shell=True)
   subprocess.call('iptables -t nat -X', shell=True)

def iptables_up(tortpdir, user):
   """ This function add iptables rules for redirect all user traffic to tortp """
   subprocess.call('iptables-save > %s/%s' % (tortpdir, "iptables.txt"), shell=True)
   # Redirect DNSTor port (9053)
   subprocess.call('iptables -t nat -A OUTPUT ! -o lo -p udp -m owner --uid-owner %s -m udp --dport 53 -j REDIRECT --to-ports 9053' % user, shell=True)
   subprocess.call('iptables -t filter -A OUTPUT -p udp -m owner --uid-owner %s -m udp --dport 53 -j ACCEPT' % user, shell=True)
   subprocess.call('iptables -t filter -A OUTPUT -p tcp -m owner --uid-owner %s -m tcp --dport 53 -j ACCEPT' % user, shell=True)
   # Redirect to Transparent Proxy Tor (9040)
   subprocess.call('iptables -t nat -A OUTPUT ! -o lo -p tcp -m owner --uid-owner %s -m tcp -j REDIRECT --to-ports 9040' % user, shell=True)
   subprocess.call('iptables -t filter -A OUTPUT -p tcp -m owner --uid-owner %s -m tcp --dport 9040 -j ACCEPT' % user, shell=True)
   subprocess.call('iptables -t filter -A OUTPUT ! -o lo -m owner --uid-owner %s -j DROP' % user, shell=True)
   subprocess.call('iptables -t nat -A OUTPUT -p tcp -m owner --uid-owner %s -m tcp --syn -d 127.0.0.1 --dport 9051 -j ACCEPT' % user, shell=True)

def iptables_down(tortpdir):
   try:
      subprocess.call('iptables-restore < %s/%s' % (tortpdir, "iptables.txt"), shell=True)
   except IOError as e:
      iptables_clean()
      print e

def resolvconf(tortpdir):
   """ Backup and modify resolv configuration file """
   try:
      copy2("/etc/resolv.conf",tortpdir)
   except IOError as e:
      print e
   resolv = open('/etc/resolv.conf', 'w')
   resolv.write('nameserver 127.0.0.1\n')
   resolv.close()

def dnsmasq(tortpdir):
   """ Backup and modify dnsmasq configuration file """
   try:
      copy2("/etc/dnsmasq.conf",tortpdir)
   except IOError as e:
      print e
   dmasq = open('/etc/dnsmasq.conf', 'w')
   dmasq.write('no-resolv\n')
   dmasq.write('server=127.0.0.1#9053\n')
   dmasq.write('listen-address=127.0.0.1\n')

def enable_tordns():
   with Controller.from_port(port = 9051) as controller:
      controller.authenticate()
      controller.set_options({"DNSPort": "9053", "AutomapHostsOnResolve": "1", "AutomapHostsSuffixes": ".exit,.onion"})

def enable_torproxy():
   with Controller.from_port(port = 9051) as controller:
      controller.authenticate()
      controller.set_options({"VirtualAddrNetwork": "10.192.0.0/10", "TransPort": "9040", "TransListenAddress": "127.0.0.1","AvoidDiskWrites": "1", "WarnUnsafeSocks": "1"})

def exit_info():
   """ Print info about my exit node """
   with Controller.from_port(port = 9051) as controller:
      controller.authenticate()
      for circ in controller.get_circuits():
         if circ.status != CircStatus.BUILT:
            continue
         exit_fp, exit_nickname = circ.path[-1]
         exit_desc = controller.get_network_status(exit_fp, None)
         exit_address = exit_desc.address if exit_desc else 'unknown'
         print "Exit relay"
         print "  fingerprint: %s" % exit_fp
         print "  nickname: %s" % exit_nickname
         print "  address: %s" % exit_address

def tor_new():
   """ Create a new tor circuit """
   try:
      stem.socket.ControlPort(port = 9051)
   except stem.SocketError as exc:
      print "Unable to connect to port 9051 (%s)" % exc
      print "Please add 'ControlPort 9051' on your /etc/tor/torrc configuration"
      sys.exit(1)
   with Controller.from_port(port = 9051) as controller:
      controller.authenticate()
      controller.signal(stem.Signal.NEWNYM)
      notify("TORtp", "New Tor circuit created")

def start(tortpdir):
   try:
      stem.socket.ControlPort(port = 9051)
   except stem.SocketError as exc:
      print "Unable to connect to port 9051 (%s)" % exc
      print "Please add 'ControlPort 9051' on your /etc/tor/torrc configuration"
      sys.exit(1)
   iptables_clean()
   iptables_up(tortpdir, check_user())
   enable_tordns()
   enable_torproxy()
   resolvconf(tortpdir)
   dnsmasq(tortpdir)
   subprocess.call('/etc/init.d/dnsmasq restart', shell=True)
   notify("TORtp", "Tor Transparent Proxy enabled")

def stop(tortpdir):
   """ Restore all original files"""
   try:
      copy2("%s/resolv.conf" % tortpdir, "/etc")
      copy2("%s/dnsmasq.conf" % tortpdir, "/etc")
   except IOError as e:
      print e
   subprocess.call('/etc/init.d/dnsmasq restart', shell=True)
   subprocess.call('/etc/init.d/tor reload', shell=True)
   iptables_down(tortpdir)
   notify("TORtp", "Tor Transparent Proxy disabled")

def main(arg):
   parser = argparse.ArgumentParser()
   parser.add_argument("do", help="start | stop | new | info")
   args = parser.parse_args()
   if args.do == "start":
      start(tortp_dir(set_home(check_user())))
   if args.do == "stop":
      stop(tortp_dir(set_home(check_user())))
   if args.do == "new":
      tor_new()
   if args.do == "info":
      exit_info()
   if args.do != "start" and args.do != "stop" and args.do != "new" and args.do != "info":
      print 'Type "tortp.py -h" for options'

if __name__ == '__main__':
   main(sys.argv[1:])


tortp-gtk.py
#!/usr/bin/env python
import pygtk
pygtk.require('2.0')
import gtk
import gobject
import gettext
import tortp

gettext.bindtextdomain('tortp-gtk', 'locale')
gettext.textdomain('tortp-gtk')
_ = gettext.gettext


class Icon(gtk.StatusIcon):

    def __init__(self):
        gtk.StatusIcon.__init__(self)
        self.set_from_file('/usr/share/pixmaps/anonymous.ico')
        self.connect("activate", self.load)
        self.connect("popup-menu", self.right_click_event)
        self.set_tooltip("TorTP")
        self.window = None
        self.load(None)

    def start(self):
        gtk.main()

    def stop(self):
        gtk.main_quit()

    def load(self, param):
        if self.window is None:
            self.window = MainWindow()
            self.window.connect("destroy", self.clear_window)

    def clear_window(self, widget, data=None):
        self.window = None

    def right_click_event(self, icon, button, time):
        menu = gtk.Menu()
        about = gtk.MenuItem(_("About"))
        quit = gtk.MenuItem(_("Quit"))

        about.connect("activate", self.show_about_dialog)
        quit.connect("activate", gtk.main_quit)

        menu.append(about)
        menu.append(quit)
        menu.show_all()
        menu.popup(None, None, gtk.status_icon_position_menu, button, time, self)

    def show_about_dialog(self, widget):
        about_dialog = gtk.AboutDialog()
        about_dialog.set_destroy_with_parent(True)
        about_dialog.set_name("TorTransparentProxy")
        about_dialog.set_website("https://github.com/vinc3nt/stem-tortp")
        about_dialog.set_version("0.1")
        about_dialog.set_authors(["vinc3nt", "paskao"])

        about_dialog.run()
        about_dialog.destroy()


class MainWindow(gtk.Window):

    def __init__(self):
        gtk.Window.__init__(self, gtk.WINDOW_TOPLEVEL)
        self.set_border_width(10)

        box = gtk.VBox(False, 0)
        notebook = gtk.Notebook()
        box.add(notebook)
        self.add(box)

        tp = TransparentProxyBox()
        label = gtk.Label(_("Transparent Proxy"))
        notebook.append_page(tp, label)

        tpinfo = TransparentProxyInfoBox()
        tpinfo.set_scroll_adjustments(None, None)
	scroll = gtk.ScrolledWindow()
        scroll.set_policy(gtk.POLICY_NEVER, gtk.POLICY_ALWAYS)
        scroll.add_with_viewport(tpinfo)
        label_info = gtk.Label(_("Infos"))
        notebook.append_page(scroll, label_info)

        self.show_all()


class TransparentProxyInfoBox(gtk.VBox):

    def __init__(self):
        gtk.VBox.__init__(self, False, 0)
        self.liststore = gtk.ListStore(str, str, str)
        self.treeview = gtk.TreeView(model=self.liststore)
        self.add(self.treeview)
        renderer_text = gtk.CellRendererText()
        column0 = gtk.TreeViewColumn(_("Fingerprint"), renderer_text, text=0)
        self.treeview.append_column(column0)
        column1 = gtk.TreeViewColumn(_("Nickname"), renderer_text, text=1)
        self.treeview.append_column(column1)
        column2 = gtk.TreeViewColumn(_("IP address"), renderer_text, text=2)
        self.treeview.append_column(column2)
        self.load_model()

    def load_model(self):
       self.liststore.clear()
       try:
           for item in tortp.get_info():
               self.liststore.append(item)
       except:
           pass
       gobject.timeout_add_seconds(5, self.load_model)


class TransparentProxyBox(gtk.VBox):

    def __init__(self):
        gtk.VBox.__init__(self, False, 0)
        self.description_text = _("""TorTP redirect all TCP ad UDP (dns) traffic generated from paranoid user in a transparent way.
After TorTP starts be sure that Tor is working looking at this page: https://check.torproject.org""")

        self.buttons_box = gtk.HButtonBox()
        self.buttons_box.set_layout(gtk.BUTTONBOX_END)
        self.description = gtk.Label(self.description_text)
        self.start_button = gtk.Button(_("Start"))
        self.change_button = gtk.Button(_("New circuit"))
        self.stop_button = gtk.Button(_("Stop"))
        self.description.set_line_wrap(True)
        self.description.set_single_line_mode(False)

        self.start_button.connect("clicked", self.start, None)
        self.change_button.connect("clicked", self.change, None)
        self.stop_button.connect("clicked", self.stop, None)

        # Add components
        self.buttons_box.add(self.start_button)
        self.buttons_box.add(self.change_button)
        self.buttons_box.add(self.stop_button)
        self.add(self.description)
        self.add(self.buttons_box)

        is_running = tortp.is_running()
        self.start_button.set_sensitive(not is_running)
        self.change_button.set_sensitive(is_running)
        self.stop_button.set_sensitive(is_running)

    def start(self, widget, data=None):
        tortp.do_start()
        self.start_button.set_sensitive(False)
        self.stop_button.set_sensitive(True)
        self.change_button.set_sensitive(True)

    def change(self, widget, data=None):
        tortp.tor_new()

    def stop(self, widget, data=None):
        tortp.do_stop()
        self.start_button.set_sensitive(True)
        self.stop_button.set_sensitive(False)
        self.change_button.set_sensitive(False)


if __name__ == "__main__":
    icon = Icon()
    icon.start()


tortp.py
#!/usr/bin/env python

"""
TorTP is a simple way to implement
Tor Transparent Proxy in our GNU/Linux Box

TorTP use Tor's control port for setup
Transparen Proxy and DNS server capability
on TOR, without override default configuration file
with a custom torrc
"""

import subprocess
import os
import stem
from stem.control import Controller
from stem.version import get_system_tor_version
import stem.util
from stem.util import system
import stem.process
from stem import CircStatus
from shutil import copy2
import sys
import pwd
import urllib
import re

def notify(title, message):
    """
    Notification system
    """
    print("[%s]: %s" % (title, message))
    return

def check_user():
   """
   Only root can do that!
   """
   uid = os.getuid()
   if uid == 0:
      return os.environ.get('SUDO_UID', 0)
   else:
      notify("TorTP", "[!] Only root can do that!")
      sys.exit(1)

def get_toruser():
    """
    Get tor username
    """
    pid = system.get_pid_by_port(9050)
    toruser = system.get_user(pid)
    return toruser

def get_home(user):
   """
   Get user home path
   """
   return pwd.getpwuid(int(user))[5]

def tortp_dir(home):
   """
   Create directory /home/$user/.tortp
   """
   tortpdir = "%s/.tortp" % home
   if not os.path.exists(tortpdir):
      os.makedirs(tortpdir)
      notify("TorTP", "[+] Directory %s created" % tortpdir)
   return tortpdir

def check_sys_dependencies():
   """
   Check if all dependencies are installed
   """
   devnull = open(os.devnull,"w")
   dnsmasq = subprocess.call(["dpkg","-s","dnsmasq"],stdout=devnull,stderr=subprocess.STDOUT)
   if dnsmasq != 0:
      notify("TorTP", "[!] Dnsmasq is not installed")
      sys.exit(1)
   tor = subprocess.call(["dpkg","-s","tor"],stdout=devnull,stderr=subprocess.STDOUT)
   if tor != 0:
      notify("TorTP", "[!] Tor is not installed")
      sys.exit(1)
   devnull.close()


def iptables_clean():
   """
   This function remove all iptables rules
   """
   subprocess.call(['iptables', '-F'])
   subprocess.call(['iptables', '-X'])
   subprocess.call(['iptables', '-t', 'nat', '-F'])
   subprocess.call(['iptables', '-t', 'nat', '-X'])

def iptables_up(tortpdir, toruser):
   """
   This function make backup and add iptables rules for redirect all network traffic to TorTP.
   Only except with debian-tor user.
   """
   ipt = open("%s/iptables.txt" % tortpdir, "w")
   subprocess.call(['iptables-save'], stdout=ipt)
   ipt.close()
   # Redirect DNSTor port (9053)
   subprocess.call(['iptables', '-t', 'nat', '-A', 'OUTPUT', '!', '-o', 'lo', '-p', 'udp', '-m', 'owner', '!', '--uid-owner', '%s' % toruser, '-m', 'udp', '--dport', '53', '-j', 'REDIRECT', '--to-ports', '9053'])
   subprocess.call(['iptables', '-t', 'filter', '-A', 'OUTPUT', '-p', 'udp', '-m', 'owner', '!', '--uid-owner', '%s' % toruser, '-m', 'udp', '--dport', '53', '-j', 'ACCEPT'])
   subprocess.call(['iptables', '-t', 'filter', '-A', 'OUTPUT', '-p', 'tcp', '-m', 'owner', '!', '--uid-owner', '%s' % toruser, '-m', 'tcp', '--dport', '53', '-j', 'ACCEPT'])
   # Redirect to Transparent Proxy Tor (9040)
   subprocess.call(['iptables', '-t', 'nat', '-A', 'OUTPUT', '!', '-o', 'lo', '-p', 'tcp', '-m', 'owner', '!', '--uid-owner', '%s' % toruser, '-m', 'tcp', '-j', 'REDIRECT', '--to-ports', '9040'])
   subprocess.call(['iptables', '-t', 'filter', '-A', 'OUTPUT', '-p', 'tcp', '-m', 'owner', '!', '--uid-owner', '%s' % toruser, '-m', 'tcp', '--dport', '9040', '-j', 'ACCEPT'])
   subprocess.call(['iptables', '-t', 'filter', '-A', 'OUTPUT', '!', '-o', 'lo', '-m', 'owner', '!', '--uid-owner', '%s' % toruser, '-j', 'DROP'])
   subprocess.call(['iptables', '-t', 'nat', '-A', 'OUTPUT', '-p', 'tcp', '-m', 'owner', '!', '--uid-owner', '%s' % toruser, '-m', 'tcp', '--syn', '-d', '127.0.0.1', '--dport', '9051', '-j', 'ACCEPT'])

def iptables_down(tortpdir):
   """
   Restore original iptables rules
   """
   try:
      subprocess.call('iptables-restore < %s/iptables.txt' % tortpdir, shell=True)
      os.remove("%s/iptables.txt" % tortpdir)
   except IOError as e:
      iptables_clean()
      print e

def resolvconf(tortpdir):
   """
   Backup and modify resolv configuration file
   """
   try:
      copy2("/etc/resolv.conf",tortpdir)
   except IOError as e:
      print e
   resolv = open('/etc/resolv.conf', 'w')
   resolv.write('nameserver 127.0.0.1\n')
   resolv.close()

def dnsmasq(tortpdir):
   """
   Backup and modify dnsmasq configuration file
   """
   try:
      copy2("/etc/dnsmasq.conf",tortpdir)
   except IOError as e:
      print e
   dmasq = open('/etc/dnsmasq.conf', 'w')
   dmasq.write('no-resolv\n')
   dmasq.write('server=127.0.0.1#9053\n')
   dmasq.write('listen-address=127.0.0.1\n')
   dmasq.close()

def myip():
   """
   Get my IP from check.torproject.org
   """
   url = "http://check.torproject.org"
   request = urllib.urlopen(url).read()
   myip = re.findall(r"\d{1,3}\.\d{1,3}\.\d{1,3}.\d{1,3}", request)
   return myip[0]

def check_tortp(myip, exit):
   """
   Check if my IP is a Tor exit node
   """
   if system.is_running("tor"):
      try:
         if myip not in exit['ipaddress']:
            notify("TorTP", "[-] Sorry. TorTP is not working: %s" % myip)
            sys.exit(1)
      except stem.SocketError as exc:
         notify("TorTP", "[!] Unable to connect to port 9051 (%s)" % exc)
         sys.exit(1)

      notify("TorTP", "[+] Congratulations. TorTP is working: %s" % myip)
      return myip
   else:
      notify("TorTP", "[!] Tor is not running")
      sys.exit(3)

def get_exit():
   """
   Get list of exit node from stem
   """
   if system.is_running("tor"):
      try:
         with Controller.from_port(port = 9051) as controller:
            controller.authenticate()
            exit = {'count': [], 'fingerprint': [], 'nickname': [], 'ipaddress': []}
            count = -1
            for circ in controller.get_circuits():
               if circ.status != CircStatus.BUILT:
                  continue
               exit_fp, exit_nickname = circ.path[-1]
               exit_desc = controller.get_network_status(exit_fp, None)
               exit_address = exit_desc.address if exit_desc else 'unknown'
               count += 1
               exit['count'].append(count)
               exit['fingerprint'].append(exit_fp)
               exit['nickname'].append(exit_nickname)
               exit['ipaddress'].append(exit_address)
         return exit
      except stem.SocketError as exc:
         notify("TorTP", "[!] Unable to connect to port 9051 (%s)" % exc)
         sys.exit(1)
   else:
      notify("TorTP", "[!] Tor is not running")


def exit_info(exit):
   """
   Print info about my exit node
   """
   torversion = get_system_tor_version()
   print "Tor version: %s\n" % torversion
   for i in exit['count']:
      print "  nickname: %s" % exit['nickname'][i]
      print "  address: %s" % exit['ipaddress'][i]
      print "  fingerprint: %s\n" % exit['fingerprint'][i]

def tor_new():
   """
   Create a new tor circuit
   """
   try:
      stem.socket.ControlPort(port = 9051)
   except stem.SocketError as exc:
      notify("TorTP", "[!] Unable to connect to port 9051 (%s)" % exc)
      sys.exit(1)
   with Controller.from_port(port = 9051) as controller:
      controller.authenticate()
      controller.signal(stem.Signal.NEWNYM)
      notify("TorTP", "[+] New Tor circuit created")

def tor_new_process():
    curr_uid = os.getuid()
    #print("Current uid: ", curr_uid)
    """
    Changing to debian-tor (E)UID and (E)GUID
    and start a new tor process
    """
    debian_tor_uid = getpwnam('debian-tor').pw_uid
    debian_tor_gid = getpwnam('debian-tor').pw_gid
    os.setgid(debian_tor_gid)
    os.setuid(debian_tor_uid)
    os.setegid(debian_tor_gid)
    os.seteuid(debian_tor_uid)
    #print debian_tor_uid
    os.environ['HOME'] = "/var/lib/tor"

    tor_process = stem.process.launch_tor_with_config(
      config = {
        'SocksPort': '6666',
        'DNSPort': '9053',
        'DNSListenAddress': '127.0.0.1',
        'AutomapHostsOnResolve': '1',
        'AutomapHostsSuffixes': '.exit,.onion',
        'VirtualAddrNetwork': '10.192.0.0/10',
        'TransPort': '9040',
        'TransListenAddress': '127.0.0.1',
        'AvoidDiskWrites': '1',
        'WarnUnsafeSocks': '1',
      })


def start(tortpdir):
   """
   Start TorTP
   """
  if os.path.exists("%s/resolv.conf" % tortpdir) and os.path.exists("%s/dnsmasq.conf" % tortpdir) and os.path.exists("%s/iptables.txt" % tortpdir):
     notify("TorTP", "[!] TorTP is already running")
     sys.exit(2)
  else:
     check_sys_dependencies()
     iptables_clean()
     iptables_up(tortpdir, get_toruser())
     resolvconf(tortpdir)
     dnsmasq(tortpdir)
     devnull = open(os.devnull,"w")
     subprocess.call(['/etc/init.d/dnsmasq', 'restart'], stdout=devnull)
     devnull.close()
     tor_new_process()
     notify("TorTP", "[+] Tor Transparent Proxy enabled")

def stop(tortpdir):
   """
   Stop TorTP and restore original network configuration
   """
   try:
      copy2("%s/resolv.conf" % tortpdir, "/etc")
      copy2("%s/dnsmasq.conf" % tortpdir, "/etc")
      os.remove("%s/resolv.conf" % tortpdir)
      os.remove("%s/dnsmasq.conf" % tortpdir)
   except IOError:
      notify("TorTP", "[!] TorTP seems already disabled")
      sys.exit(1)
   devnull = open(os.devnull,"w")
   subprocess.call(['/etc/init.d/dnsmasq', 'restart'], stdout=devnull)
   devnull.close()
   iptables_down(tortpdir)
   notify("TorTP", "[+] Tor Transparent Proxy disabled")

def is_running():
   """
   check if TorTP is running
   """
   #TODO: change check
   path = tortp_dir(get_home(check_user()))
   file_path = os.path.join(path, "resolv.conf")
   return os.path.exists(file_path)

def do_start():
   start(tortp_dir(get_home(check_user())))

def do_stop():
   stop(tortp_dir(get_home(check_user())))

def do_check():
   return check_tortp(myip(), get_exit())
